File: packages/types/src/social.ts
import { z } from 'zod';
import { IsoDateTimeSchema, UuidSchema } from './shared';

export const PostCategorySchema = z.enum(['events', 'news', 'announcements', 'community']);
export type PostCategory = z.infer<typeof PostCategorySchema>;

export const PostSlideSchema = z.object({
  bg: z.string(),
  emoji: z.string().optional(),
  title: z.string(),
  sub: z.string().optional(),
});
export type PostSlide = z.infer<typeof PostSlideSchema>;

export const PostSchema = z.object({
  id: UuidSchema,
  category: PostCategorySchema,
  caption: z.string(),
  slides: z.array(PostSlideSchema),
  isPinned: z.boolean(),
  authorId: UuidSchema.nullable(),
  publishedAt: IsoDateTimeSchema,
  createdAt: IsoDateTimeSchema,
});
export type Post = z.infer<typeof PostSchema>;

export const ReelVisualKindSchema = z.enum(['water', 'zen', 'fire', 'leaves', 'sparkle']);
export type ReelVisualKind = z.infer<typeof ReelVisualKindSchema>;

export const ReelSchema = z.object({
  id: UuidSchema,
  category: z.string(),
  title: z.string(),
  description: z.string().nullable(),
  visualKind: ReelVisualKindSchema.nullable(),
  videoUrl: z.string().url().nullable(),
  authorId: UuidSchema.nullable(),
  publishedAt: IsoDateTimeSchema,
});
export type Reel = z.infer<typeof ReelSchema>;

export const StorySchema = z.object({
  id: UuidSchema,
  emoji: z.string().nullable(),
  label: z.string(),
  title: z.string().nullable(),
  subtitle: z.string().nullable(),
  visual: z.string().nullable(),
  expiresAt: IsoDateTimeSchema.nullable(),
  authorId: UuidSchema.nullable(),
  createdAt: IsoDateTimeSchema,
});
export type Story = z.infer<typeof StorySchema>;

export const CommentSchema = z.object({
  id: UuidSchema,
  postId: UuidSchema,
  userId: UuidSchema,
  content: z.string(),
  createdAt: IsoDateTimeSchema,
});
export type Comment = z.infer<typeof CommentSchema>;

File: apps/api/app/models/social.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, CreatedAt, UuidPk, utcnow


class Post(UuidPk, CreatedAt, Base):
    __tablename__ = "posts"

    category: Mapped[str] = mapped_column(String, nullable=False, index=True)
    caption: Mapped[str] = mapped_column(Text, nullable=False)
    slides: Mapped[list[dict]] = mapped_column(JSONB, nullable=False, default=list)
    is_pinned: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    author_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    published_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, nullable=False
    )


class Reel(UuidPk, Base):
    __tablename__ = "reels"

    category: Mapped[str] = mapped_column(String, nullable=False, index=True)
    title: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    visual_kind: Mapped[str | None] = mapped_column(String, nullable=True)
    video_url: Mapped[str | None] = mapped_column(String, nullable=True)
    author_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    published_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, nullable=False
    )


class Story(UuidPk, CreatedAt, Base):
    __tablename__ = "stories"

    emoji: Mapped[str | None] = mapped_column(String, nullable=True)
    label: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str | None] = mapped_column(String, nullable=True)
    subtitle: Mapped[str | None] = mapped_column(String, nullable=True)
    visual: Mapped[str | None] = mapped_column(String, nullable=True)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    author_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )


class Comment(UuidPk, CreatedAt, Base):
    __tablename__ = "comments"

    post_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("posts.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
